pkg/util/awsutil/s3factory/client_test.go

```diff
@@ -38,3 +38,45 @@ func TestSetupAWSConfig(t *testing.T) {
 		t.Errorf("got: %s wanted: %s", *opts.Config.Endpoint, e)
 	}
 }
+
+func TestS3ClientForIAMRoles(t *testing.T) {
+	tests := []struct {
+		name   string
+		cfg    ClientConfig
+		expErr bool
+	}{
+		{
+			name: "A configuration without AWS secret, should return a client without credentials.",
+			cfg: ClientConfig{
+				Endpoint: "s3.eu-west-1.amazonaws.com",
+			},
+			expErr: false,
+		},
+		{
+			name: "A configuration without AWS secret and bad endpoint without region, should return a error.",
+			cfg: ClientConfig{
+				Endpoint: "s3.amazonaws.com",
```

Comment:
Curious why this is considered a bad endpoint? This is the default endpoint in the SDK, as an S3 user I'd expect it to be valid.

Reply:
Although S3 is multi-region, you need to specify the region, so we don't need to add another option like the `region` (this way improving the configuration usage experience), and you provide directly the region on the url (is a common practice). The [error](https://github.com/coreos/etcd-operator/pull/2021/files#diff-cc5b6d77b89214ec48aaee598700dd5bR78) is self-descriptive to let the user know what's happening.

Reply:
This may be out of scope for this PR but having the default endpoint cause an error is counter intuitive. Why do you need to specify the region if S3 is multi-region and s3.amazonaws.com is valid? It is not required to pass in an endpoint or a region when constructing a client.

Ideally the logic would support using straight defaults for everything or an optional configured endpoint and if that endpoint specifies a region then also the region. 

Maybe one solution is to use the IsAmazonEndpoint function in that minio utility. If that is true set the region to us-east-1?

Reply:
Replication on different AWS regions could be hours. In that case, a random default zone could be failing for hours although you know what's your main region.

And a region is required always for clients, another thing is that people have on env vars, or profiles their default zones at the moment of executing applications: https://github.com/aws/aws-sdk-go/blob/master/aws/config.go#L48-L49

Reply:
I agree with @danbeaulieu on this one. Making the default fail doesn't make sense. I should be able to configure this without specifying an endpoint.